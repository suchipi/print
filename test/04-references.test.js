const print = require("../print");

describe("Object references", () => {
	it("refers to input using a label", () => {
		const foo = { bar: "Baz" };
		expect(print(foo, "foo")).toMatchSnapshot();
	});

	it("indicates references using `->`", () => {
		const input = {};
		input.self = input;
		expect(print(input, "input")).toMatchSnapshot();

		input.foo = {};
		input.bar = [];
		input.baz = input.foo;
		expect(print(input, "input")).toMatchSnapshot();
		input.bar.push(input.baz, (input.qux = {}));
		expect(print(input, "input")).toMatchSnapshot();
		input.qux = input.bar[1].obj = [];
		expect(print(input, "input")).toMatchSnapshot();
	});

	it("defaults to the label `{root}`", () => {
		const foo = {};
		foo.bar = foo;
		expect(print(foo)).toMatchSnapshot();
	});

	it("prints symbol-type labels", () => {
		const foo = { bar: "baz" };
		const sym = Symbol("Foo");
		expect(print(foo, sym)).toMatchSnapshot();
		foo[sym] = foo;
		expect(print(foo, sym)).toMatchSnapshot();
		expect(print(foo, Symbol.iterator)).toMatchSnapshot();

		const bar = Symbol("Bar");
		const qux = Symbol("Qux");
		foo[bar] = [{}];
		foo[qux] = foo[bar][0];
		expect(print(foo, "input")).toMatchSnapshot();
		foo[bar][0].qul = { name: "Quul" };
		foo[qux] = foo[bar][0].qul;
		expect(print(foo, sym)).toMatchSnapshot();
	});
});
