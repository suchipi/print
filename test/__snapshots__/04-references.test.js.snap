// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object references defaults to the label \`{root}\` 1`] = `
{
	bar: -> {root}
}
`;

exports[`Object references indicates references using \`->\` 1`] = `
input: {
	self: -> input
}
`;

exports[`Object references indicates references using \`->\` 2`] = `
input: {
	self: -> input
	foo: {}
	bar: []
	baz: -> input.foo
}
`;

exports[`Object references indicates references using \`->\` 3`] = `
input: {
	self: -> input
	foo: {}
	bar: [
		-> input.foo
		{}
	]
	baz: -> input.foo
	qux: -> input.bar[1]
}
`;

exports[`Object references indicates references using \`->\` 4`] = `
input: {
	self: -> input
	foo: {}
	bar: [
		-> input.foo
		{
			obj: []
		}
	]
	baz: -> input.foo
	qux: -> input.bar[1].obj
}
`;

exports[`Object references prints symbol-type labels 1`] = `
Symbol(Foo): {
	bar: "baz"
}
`;

exports[`Object references prints symbol-type labels 2`] = `
Symbol(Foo): {
	bar: "baz"
	Symbol(Foo): -> Symbol(Foo)
}
`;

exports[`Object references prints symbol-type labels 3`] = `
@@iterator: {
	bar: "baz"
	Symbol(Foo): -> @@iterator
}
`;

exports[`Object references prints symbol-type labels 4`] = `
input: {
	bar: "baz"
	Symbol(Foo): -> input
	Symbol(Bar): [
		{}
	]
	Symbol(Qux): -> input.Symbol(Bar)[0]
}
`;

exports[`Object references prints symbol-type labels 5`] = `
Symbol(Foo): {
	bar: "baz"
	Symbol(Foo): -> Symbol(Foo)
	Symbol(Bar): [
		{
			qul: {
				name: "Quul"
			}
		}
	]
	Symbol(Qux): -> Symbol(Foo).Symbol(Bar)[0].qul
}
`;

exports[`Object references refers to input using a label 1`] = `
foo: {
	bar: "Baz"
}
`;
